src/extra/components/adq.py

```diff
@@ -0,0 +1,1294 @@
+
+from extra_data.read_machinery import roi_shape
+from .pulses import XrayPulses
+from .utils import _isinstance_no_import
+from ._adq import _reshape_flat_pulses
+
+
+class AdqChannel:
+    """A high-level interface to ADQ digitizer channels.
+
+    Fast ADQ digitizers are used to acquire analog signals at GHz sample
+    rates. In addition to the immediate raw data, their FPGA cores are
+    able to run analyss algorithms like zero suppression or peak
+    integration during acqusition.
+
+    This component is able to access the instrument data saved with
+    these devices in various representations and data formats. All
+    processing performed implicitly by the retrieval methods is also
+    available to run explicitly on a set of data.
+
+    Currently, only access to raw data is available either by train or
+    pulse as well as digitized via edge discriminators.
+
+    Args:
+        data (extra.data.DataCollection): Data to access digitizer from.
+        channel (str): Channel name with or without underscore, e.g.
+            1_A or 3C.
+        digitizer (str, optional): Source name of an ADQ digitizer, only
+            needed if the data includes more than one such device or
+            none could be detected automatically.
+        interleaved (bool, optional): Whether this digitizer channel was
+            interleaving samples or not, only needed if it could not be
+            detected automatically.
+        clock_ratio (int, optional): Digitizer sampling clock as
+            multiple of the bunch repetition rate (4.5 MHz).
+        pulses (extra.components.pulses.PulsePattern): Pulse component
+            to pull pulse information, by default a
+            [XrayPulses][extra.components.XrayPulses] object is
+            constructed unless pulse information are explicitly disabled
+            by passing `False`. Most pulse separation operations require
+            this to be present.
+        first_pulse_offset (int, optional): Sample where the first
+            pulse begins, 10000 by default.
+        single_pulse_length (int, optional): Samples per pulse for
+            the case all trains only contain a single pulse.
+        preprocess_kw (dict, optional): Custom settings for raw data
+            preprocessing done by this component, please see arguments
+            of [AdqChannel.preprocess].
+    """
+
+    # 4.5 Mhz, see extra.components.pulses.
+    _bunch_repetition_rate = 1.3e9 / 288
+
+    _adq_pipeline_re = re.compile(r'(\w+)\/ADC\/(\d):network')
+    _adq_channel_re = re.compile(
+        r'digitizers.channel_(\d)_([A-Z]).raw.samples')
+
+    # Set of digitizers using 3G boards with a 392 clock ratio that do
+    # not advertise as such.
+    _3g_digitizer = {'SQS_DIGITIZER_UTC2/ADC/1'}
+
+    def __init__(self, data, channel, digitizer=None, pulses=None,
+                 first_pulse_offset=10000, single_pulse_length=25000,
+                 interleaved=None, clock_ratio=None, preprocess_kw={}):
```

Comment:
I moved `first_pulse_offset`, `single_pulse_length, `preprocess_kw` from individual methods to the initializer to simplify their signature aind since I expect them to a) change rearely and b) only by experts.